odxtools/diaglayer.py

```diff
@@ -699,6 +594,126 @@ def get_tester_present_time(self, protocol_name: Optional[str] = None) -> Option
+
+        Example:
+        Let there be four services with corresponding requests:
+        * Request 1 has the coded constant prefix `12 34`.
+        * Request 2 has the coded constant prefix `12 34`.
+        * Request 3 has the coded constant prefix `12 56`.
+        * Request 4 has the coded constant prefix `12 56 00`.
+
+        Then, the constructed prefix tree is the dict
+        ```
+        {0x12: {0x34: {-1: [<Service 1>, <Service 2>]},
+                0x56: {-1: [<Service 3>],
+                       0x0: {-1: [<Service 4>]}
+                       }}}
+        ```
+        Note, that the inner `-1` are constant to distinguish them from possible service IDs.
+
+        Also note, that it is actually allowed that
+        (a) SIDs for different services are the same like for service 1 and 2 (thus each leaf node is a list) and
+        (b) one SID is the prefix of another SID like for service 3 and 4 (thus the constant `-1` key).
+        """
+        services = [s for s in self._services if isinstance(s, DiagService)]
+        prefix_tree = {}
+        for s in services:
+            # Compute prefixes for the request and all responses
+            request_prefix = s.request.coded_const_prefix()
+            prefixes = [request_prefix] + [
+                message.coded_const_prefix(request_prefix=request_prefix)
+                for message in chain(s.positive_responses, s.negative_responses)
+            ]
+            for coded_prefix in prefixes:
+                # Traverse prefix tree
+                sub_tree = prefix_tree
+                for b in coded_prefix:
+                    if sub_tree.get(b) is None:
+                        sub_tree[b] = {}
+                    sub_tree = sub_tree.get(b)
+
+                    assert isinstance(
+                        sub_tree,
+                        dict), f"{sub_tree} has type {type(sub_tree)}. How did this happen?"
+                # Add service as leaf to prefix tree
+                if sub_tree.get(-1) is None:
+                    sub_tree[-1] = [s]
+                else:
+                    sub_tree[-1].append(s)
+        return prefix_tree
+
+    def _find_services_for_uds(self, message: Union[bytes, bytearray]):
+        if not hasattr(self, "_prefix_tree"):
+            # Compute the prefix tree the first time this decode function is called.
+            self._prefix_tree = self._build_coded_prefix_tree()
+        prefix_tree = self._prefix_tree
+
+        # Find matching service(s) in prefix tree
+        possible_services = []
+        for b in message:
+            if prefix_tree.get(b) is not None:
+                assert isinstance(prefix_tree.get(b), dict)
+                prefix_tree = prefix_tree.get(b)
+            else:
+                break
+            if -1 in prefix_tree:
+                possible_services += prefix_tree[-1]
+        return possible_services
+
+    def decode(self, message: Union[bytes, bytearray]) -> Iterable[Message]:
```

Comment:
just a suggestion, maybe move this function and related code to its own class, to reduce the code complexity of the diag layer

Reply:
good point. this might be stuff for a future PR...